### PR TITLE
Dvc 2584 fetch config

### PR DESCRIPTION
**Pull request by emmawillis**
Adds a config manager to handle storing and re-fetching the config. Note - I'm working on unit tests for EnvironmentConfigManager but figured I'd put this out to avoid blocking anyone who needs it

### CODE REVIEW

File: DevCycle/EnvironmentConfigManager/EnvironmentConfigManager.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        private void SetConfig(IRestResponse res)
+        {
+            if (res.StatusCode == System.Net.HttpStatusCode.NotModified)
+            {
+                //TODO log (`Config not modified, using cache, etag: ${ this.configEtag}`)
+            }
+            else if (res.StatusCode == System.Net.HttpStatusCode.OK)
+            {
+                if (configString == null && onInitialized != null) onInitialized.OnSuccess();
+                configString = res.Content;
```
Comment (ajwootto) [inline DevCycle/EnvironmentConfigManager/EnvironmentConfigManager.cs:69]: wouldnt you want to set the configString before calling onInitialized.onSuccess?

File: DevCycle/Api/DVCClient.cs [hunk 2]
```diff
@@ -160,23 +165,6 @@ public void Dispose()
             ((IDisposable)apiClient).Dispose();
         }
 
-        public async void OnClientInitialized(IDVCCallback callback)
```
Comment (squarepegg) [inline DevCycle/Api/DVCClient.cs:163]: Looks like the OnClientInitialized has been removed?

File: DevCycle/EnvironmentConfigManager/EnvironmentConfigManager.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        private string configString;
+        private string configEtag;
+
+        public EnvironmentConfigManager(string environmmentKey, DVCOptions dvcOptions = null, IDVCCallback onInitialized = null)
+        {
+            if (dvcOptions == null) dvcOptions = new DVCOptions();
+
+            this.environmmentKey = environmmentKey;
+            // TODO create logger
+            pollingIntervalMS = dvcOptions.configPollingIntervalMS >= 1000
```
Comment (squarepegg) [inline DevCycle/EnvironmentConfigManager/EnvironmentConfigManager.cs:33]: Might be clearer to put 1000 into a variable

File: DevCycle/Api/DVCClient.cs (hunk 2 above)
Comment (emmawillis) [reply DevCycle/Api/DVCClient.cs:163]: Yeah, that was the weird wrapper around the EnvironmentConfigManager's initial config fetch (where we had that public Task variable getting passed around). By just passing the OnInitialized callback to the config manager in the constructor it avoids needing a separate function to call it

File: DevCycle/EnvironmentConfigManager/EnvironmentConfigManager.cs (hunk 3 above)
Comment (emmawillis) [reply DevCycle/EnvironmentConfigManager/EnvironmentConfigManager.cs:33]: true, will do!

File: DevCycle/EnvironmentConfigManager/EnvironmentConfigManager.cs (hunk 1 above)
Comment (emmawillis) [reply DevCycle/EnvironmentConfigManager/EnvironmentConfigManager.cs:69]: oh true, fixing now

File: DevCycle.MSTests/EnvironmentConfigManagerTest.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+{
+    [TestClass]
+    public class EnvironmentConfigManagerTest
+    {
+        [TestMethod]
+        public async Task OnInitializedCallbackTest()
+        {
+            Mock<IDVCCallback> mockCallback = new Mock<IDVCCallback>();
+
+            EnvironmentConfigManager configManager = new EnvironmentConfigManager("server-6fb598fd-aa34-439d-a41b-44deed51dbf0", null, mockCallback.Object); // TODO mock rest client so dont need real server key
```
Comment (squarepegg) [inline DevCycle.MSTests/EnvironmentConfigManagerTest.cs:20]: Is this a real key? May need to delete pr / branch to ensure this key isn't in the github history.

File: DevCycle/Api/DVCClient.cs (hunk 2 above)
Comment (squarepegg) [reply DevCycle/Api/DVCClient.cs:163]: Gotcha

File: DevCycle/EnvironmentConfigManager/EnvironmentConfigManager.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                configEtag = new List<string>(headerValues).FirstOrDefault();
+            }
+            else if (configString != null)
+            {
+                //TODO log (`Failed to download config, using cached version. url: ${ url}.`)
+            }
+            else
+            {
+                if (onInitialized != null) onInitialized.OnError(new DVCException(HttpStatusCode.InternalServerError, new ErrorResponse("Failed to download DevCycle config.")));
+                else throw new DVCException(HttpStatusCode.InternalServerError, new ErrorResponse("Failed to download DevCycle config."));
```
Comment (squarepegg) [inline DevCycle/EnvironmentConfigManager/EnvironmentConfigManager.cs:83]: This exception would need to be handled in DVCClient to ensure default variables can be used even if the config could not be fetched.

File: DevCycle.MSTests/EnvironmentConfigManagerTest.cs (hunk 4 above)
Comment (emmawillis) [reply DevCycle.MSTests/EnvironmentConfigManagerTest.cs:20]: oh yikes I meant to omit all my test files, i'll delete the PR and amend the commit to remove it

File: DevCycle/EnvironmentConfigManager/EnvironmentConfigManager.cs (hunk 5 above)
Comment (squarepegg) [reply DevCycle/EnvironmentConfigManager/EnvironmentConfigManager.cs:83]: This could be handled in another ticket though.

**Pull request closed without merging**